[WPE] frameDisplayed may be called after View has been deleted

yury-s opened the pull request:
#### 7d464f717df9050f11fe11af583b82cd7800edc4
<pre>
[WPE] frameDisplayed may be called after View has been deleted
<a href="https://bugs.webkit.org/show_bug.cgi?id=264360">https://bugs.webkit.org/show_bug.cgi?id=264360</a>

Reviewed by Michael Catanzaro.

WKWPE::View is set as a client to wpe_view_backend in the view&apos;s
constructor. Since the backend may live longer than the view,
we reset clear the client in the view destuctor to avoid use
after free similar to the one mentioned in the bug.

* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::~View):

Canonical link: <a href="https://commits.webkit.org/270493@main">https://commits.webkit.org/270493@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/319a19b4c5d2ce2651235b18ff0cb9839f17d685

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4255 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26933 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27750 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/23495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6003 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1690 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25899 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3171 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/22106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28330 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23408 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23425 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26992 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/2819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4195 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6152 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/3264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/3135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

yury-s commented on Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
```diff
@@ -319,6 +319,9 @@ View::View(struct wpe_view_backend* backend, const API::PageConfiguration& baseC
 View::~View()
 {
+    wpe_view_backend_set_backend_client(m_backend, nullptr, nullptr);
+    wpe_view_backend_set_input_client(m_backend, nullptr, nullptr);
```
I would also call `wpe_view_backend_set_fullscreen_client(m_backend, nullptr, nullptr);` here, but it triggers an [assert in WPE](https://github.com/WebPlatformForEmbedded/libwpe/blob/d7c669ca6f5ec0d544c264016d270669b336c931/src/view-backend.c#L106) which needs to be addressed first:

```
    [pid=939644][err] MiniBrowser: /root/webkit/WebKitBuild/DependenciesWPE/Source/libwpe-1.14.1/src/view-backend.c:103: wpe_view_backend_set_fullscreen_client: Assertion `!backend->fullscreen_client' failed.
```

Sent a PR to WPE to relax the assert: https://github.com/WebPlatformForEmbedded/libwpe/pull/129

mcatanzaro replied:
Heh, you anticipated my review feedback. I think we should just remove that assert from libwpe. Can you report a libwpe bug for this please? Bonus points for a pull request.

yury-s replied:
I uploaded [a PR](https://github.com/WebPlatformForEmbedded/libwpe/pull/129) to libwpe, waiting for review. Do you think it's better to just remove the assert there instead of relaxing it? Other client setters don't have any asserts, so I was not sure if that was intentional or just for historical reasons.

mcatanzaro commented on Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
```diff
@@ -319,6 +319,9 @@ View::View(struct wpe_view_backend* backend, const API::PageConfiguration& baseC
 
 View::~View()
 {
+    wpe_view_backend_set_backend_client(m_backend, nullptr, nullptr);
+    wpe_view_backend_set_input_client(m_backend, nullptr, nullptr);
```
You can do something like:

```suggestion
    wpe_view_backend_set_backend_client(m_backend, nullptr, nullptr);
    wpe_view_backend_set_input_client(m_backend, nullptr, nullptr);
    // Although the fullscreen client is used for libwpe 1.11.1 and newer, we cannot
    // unregister it prior to 1.15.2 (see https://github.com/WebPlatformForEmbedded/libwpe/pull/129).
#if ENABLE(FULLSCREEN_API) && WPE_CHECK_VERSION(1, 15, 2)
    wpe_view_backend_set_fullscreen_client(m_backend, nullptr, nullptr);
#endif
```

This way we get the code correct in WebKit, but without introducing crashes for current versions of libwpe. (We'll need to make sure that assertion really does get removed in the next release.)

yury-s replied:
I'll wait for feedback on [the PR ](https://github.com/WebPlatformForEmbedded/libwpe/pull/129) and update this code if that patch is merged.

aperezdc replied:
@yury-s FYI, I agree with the change in libwpe, I want to have the change there merged, and it will be in 1.15.2, so feel free to start updating this patch as well. This one also LGTM with @mcatanzaro's suggestion applied.

yury-s replied:
Update this PR with suggested changes. PTAL.

mcatanzaro reviewed (commented):
I think this is correct. We don't create or own the wpe_view_backend and so should not destroy it. Well, WebKitWebViewBackend owns it, but WPEView does not. I think unregistering the callbacks is the right thing to do here.

mcatanzaro reviewed (approved):
Changes look good. Thanks!

Note that merge queue will refuse to process this pull request because it has more than one commit. It's not smart enough to squash them together, so you'll need to do that manually.

yury-s committed:
[WPE] frameDisplayed may be called after View has been deleted
https://bugs.webkit.org/show_bug.cgi?id=264360

Reviewed by Michael Catanzaro.

WKWPE::View is set as a client to wpe_view_backend in the view's
constructor. Since the backend may live longer than the view,
we reset clear the client in the view destuctor to avoid use
after free similar to the one mentioned in the bug.

* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::~View):

Canonical link: https://commits.webkit.org/270493@main
